File: Data Structures & Algorithms/house-robber-ii/submission-4.java
class Solution {
    public int robHelper(int[] nums, int left, int right){
        // base case
        if(left==right){
            return nums[left];
        }
        // dp
        int prev2 = nums[left];
        int prev1 = Math.max(nums[left], nums[left+1]);

        for(int i = left + 2;i<=right;i++){
            int cur = Math.max(prev2+nums[i], prev1);
            prev2 = prev1;
            prev1 = cur;
        }

        return prev1;

    }
    public int rob(int[] nums) {

        int n = nums.length;

        if(n==0){
            return 0;
        }
        if(n==1){
            return nums[0];
        }

        return Math.max(robHelper(nums, 0, n-2), robHelper(nums, 1, n-1));

        
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [1] * len(nums)
        res = 1
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], 1 + dp[j])
            res = max(dp[i], res)
        return res

        